code/modules/atmospherics/machinery/components/unary_devices/cryo.dm

```diff
@@ -11,34 +11,48 @@
 	mouse_opacity = MOUSE_OPACITY_TRANSPARENT
 	pixel_y = 22
 	appearance_flags = KEEP_TOGETHER
+	/// The current occupant being presented
+	var/mob/living/occupant
 
-/atom/movable/visual/cryo_occupant/Initialize()
+/atom/movable/visual/cryo_occupant/Initialize(mapload, obj/machinery/atmospherics/components/unary/cryo_cell/parent)
 	. = ..()
 	// Alpha masking
 	// It will follow this as the animation goes, but that's no problem as the "mask" icon state
 	// already accounts for this.
 	add_filter("alpha_mask", 1, list("type" = "alpha", "icon" = icon('icons/obj/cryogenics.dmi', "mask"), "y" = -22))
+	RegisterSignal(parent, COMSIG_MACHINERY_SET_OCCUPANT, .proc/on_set_occupant)
+	RegisterSignal(parent, COMSIG_CRYO_SET_ON, .proc/on_set_on)
+
```

Comment:
```suggestion
	RegisterSignal(parent, COMSIG_CRYO_SET_ON, .proc/on_set_on)
	RegisterSignal(parent, COMSIG_PARENT_QDELETING, .proc/on_parent_destruction)

```
Register to the parent's destruction and handle it appropriately, basically qdeleting self.
You could leave it as it is and it would soft-delete due to all the references being removed, but this offers us higher control and better debugging.

Reply:
So I should remove `QDEL_NULL(occupant_vis)` from parent destruction and put it into this callback?
Should I only set occupant to null or should I remove traits as if the occupant was being removed from the machine?

Reply:
Actually if it's handled in the parent already there's no need for this.